hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java

```diff
@@ -0,0 +1,282 @@
+					CriteriaQuery<Long> cr = cb.createQuery( Long.class );
+					Root<EntityOfBasics> root = cr.from( EntityOfBasics.class );
+
+					JpaWindow window = cb.createWindow()
+							.partitionBy( root.get( "theInt" ) )
+							.orderBy( cb.asc( root.get( "id" ) ) );
+					JpaExpression<Long> rank = cb.rank( window );
+
+					cr.select( rank ).orderBy( cb.asc( cb.literal( 1 ) ) );
+					List<Long> resultList = session.createQuery( cr ).getResultList();
+					assertEquals( 5, resultList.size() );
+					assertEquals( 1L, resultList.get( 0 ) );
+					assertEquals( 1L, resultList.get( 1 ) );
+					assertEquals( 1L, resultList.get( 2 ) );
+					assertEquals( 1L, resultList.get( 3 ) );
+					assertEquals( 2L, resultList.get( 4 ) );
+				}
+		);
+	}
+
+	@Test
+	public void testReusableWindow(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					HibernateCriteriaBuilder cb = session.getCriteriaBuilder();
+					CriteriaQuery<Tuple> cr = cb.createTupleQuery();
+					Root<EntityOfBasics> root = cr.from( EntityOfBasics.class );
+
+					JpaWindow window = cb.createWindow()
+							.partitionBy( root.get( "theInt" ) )
+							.orderBy( cb.asc( root.get( "id" ) ) );
+					JpaExpression<Long> rowNumber = cb.rowNumber( window );
+					JpaExpression<Double> percentRank = cb.percentRank( window );
+					JpaExpression<Double> cumeDist = cb.cumeDist( window );
+
+					cr.multiselect( rowNumber, percentRank, cumeDist ).orderBy( cb.asc( cb.literal( 1 ) ) );
+					List<Tuple> resultList = session.createQuery( cr ).getResultList();
+					assertEquals( 5, resultList.size() );
+					assertEquals( 0D, resultList.get( 0 ).get( 1 ) );
+					assertEquals( 0D, resultList.get( 1 ).get( 1 ) );
+					assertEquals( 0D, resultList.get( 2 ).get( 1 ) );
+					assertEquals( 0D, resultList.get( 3 ).get( 1 ) );
+					assertEquals( 1D, resultList.get( 4 ).get( 1 ) );
+					assertEquals( 1D, resultList.get( 4 ).get( 2 ) );
+				}
+		);
+	}
+
+	//	@Test
+	public void testSumWithFilterAsWindowFunction(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					// todo marco : add filter clause predicate to CriteriaBuilder ?
+					//  problem with getting the window functions as aggregate, @see SqmCriteriaNodeBuilder#windowFunction
+					TypedQuery<Long> q = session.createQuery(
+							"select sum(eob.theInt) filter (where eob.theInt > 5) over (order by eob.theInt) from EntityOfBasics eob order by eob.theInt",
+							Long.class
+					);
+					List<Long> resultList = q.getResultList();
+					assertEquals( 5L, resultList.size() );
+					assertNull( resultList.get( 0 ) );
+					assertNull( resultList.get( 1 ) );
+					assertEquals( 6L, resultList.get( 2 ) );
+					assertEquals( 13L, resultList.get( 3 ) );
+					assertEquals( 26L, resultList.get( 4 ) );
+				}
+		);
+	}
+
+	//	@Test
+	public void testFrame(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					// todo marco : how should we do window function frames in Criteria ?
```

Comment:
Just an idea for an API, something like this:

```java
jpaWindow.frameRows( cb.frameBetweenPreceeding( 2 ), cb.frameCurrentRow() );
jpaWindow.frameExclude( FrameExclusion.CURRENT_ROW )
```

with variants `frameRange` and `frameGroups` to control the frame mode. The first argument is the "start" and the second argument the "end" expression of the frame.

This way, the frame is set as a whole (in accordance to the JPA Criteria API style), yet it is still somewhat readable and typesafe. I guess it might be nice to also expose this frame information through the `JpaWindow` contract like e.g.:

```java
interface JpaWindowFrame {
    FrameKind getKind();
    Expression getFrameExpression();
}
```

Reply:
Great work. I think this is as nice as this can get. Any opinions @sebersole @dreab8 ?

Reply:
TY so much @beikov @dreab8, couldn't have done it without your help :smile:

Just pushing last changes for a couple edge-case errors for some dbs then I should be done.